src/Processors/QueryPlan/Optimizations/optimizeJoinInOrder.cpp

```diff
@@ -0,0 +1,258 @@
+
+    auto insert_pre_step = [&nodes, &node](size_t idx, auto step)
+    {
+        auto & sort_node = nodes.emplace_back();
+        sort_node.step = std::move(step);
+        sort_node.children.push_back(node.children[idx]);
+        node.children[idx] = &sort_node;
+    };
+
+    auto join_kind = join_ptr->getTableJoin().kind();
+    bool kind_allows_filtering = isInner(join_kind) || isLeft(join_kind) || isRight(join_kind);
+
+    auto has_non_const = [](const Block & block, const auto & keys)
+    {
+        for (const auto & key : keys)
+        {
+            const auto & column = block.getByName(key).column;
+            if (column && !isColumnConst(*column))
+                return true;
+        }
+        return false;
+    };
+
+    const auto & left_stream = node.children[0]->step->getOutputStream();
+    const auto & right_stream = node.children[1]->step->getOutputStream();
+
+    /// This optimization relies on the sorting that should buffer the whole stream before emitting any rows.
+    /// It doesn't hold such a guarantee for streams with const keys.
+    /// Note: it's also doesn't work with the read-in-order optimization.
+    /// No checks here because read in order is not applied if we have `CreateSetAndFilterOnTheFlyStep` in the pipeline between the reading and sorting steps.
+    bool has_non_const_keys = has_non_const(left_stream.header, join_ptr->getKeyNames(JoinTableSide::Left))
+        && has_non_const(right_stream.header, join_ptr->getKeyNames(JoinTableSide::Right));
+
+    size_t max_rows_in_filter_set = optimization_settings.max_rows_in_set_to_optimize_join;
+    if (!is_read_in_order_optimized && max_rows_in_filter_set > 0 && kind_allows_filtering && has_non_const_keys)
+    {
+        auto crosswise_connection = CreateSetAndFilterOnTheFlyStep::createCrossConnection();
+        auto add_create_set = [&](const DataStream & data_stream, const Names & key_names, JoinTableSide join_pos)
+        {
+            auto creating_set_step = std::make_unique<CreateSetAndFilterOnTheFlyStep>(
+                data_stream, key_names, max_rows_in_filter_set, crosswise_connection, join_pos);
+            creating_set_step->setStepDescription(fmt::format("Create set and filter {} joined stream", join_pos));
+
+            auto * step_raw_ptr = creating_set_step.get();
+            insert_pre_step(join_pos == JoinTableSide::Left ? 0 : 1, std::move(creating_set_step));
+            return step_raw_ptr;
+        };
+
+        auto * left_set = add_create_set(left_stream, join_ptr->getKeyNames(JoinTableSide::Left), JoinTableSide::Left);
+        auto * right_set = add_create_set(right_stream, join_ptr->getKeyNames(JoinTableSide::Right), JoinTableSide::Right);
+        if (isInnerOrLeft(join_kind))
+            right_set->setFiltering(left_set->getSet());
+        if (isInnerOrRight(join_kind))
+            left_set->setFiltering(right_set->getSet());
+    }
+
+    for (size_t i = 0; i < 2; ++i)
```

Comment:
It seems we shouldn't create sorting steps if reading in order was applied. Shouldn't it be moved under `if (!is_read_in_order_optimized ... ` statement

Reply:
Please add a comment regarding it